### PR TITLE
User can't register with the same email address by changing uppercase/lowercase letter of email

### LINKED ISSUE #28: User can register with the same email address by changing uppercase/lowercase letter of email
## Issue

- The user can register with the same email but with some capital letter

https://user-images.githubusercontent.com/1772999/108804987-69e28680-75d1-11eb-80e6-8356bff6f984.mov

- The user cannot log in with the same email without the capital letter

https://user-images.githubusercontent.com/1772999/108804942-533c2f80-75d1-11eb-9a35-71cd0e6530e9.mov

 
## Expected

Email addresses should not be case sensitive so
- User should not be able to register with the same email but with some capital letters
- User should be able to log in with the same email but with any capital letter
 
## Steps to reproduce

- The user can register with the same email but with some capital letter
1. Register with an email
2. Logout
3. Register with the same email but with some capital letter

- The user cannot log in with the same email without the capital letter
1. Register with an email with some capital letter
2. Logout
3. Login with the same email but without the capital letter

### PR DESCRIPTION
**Pull request by junan**
https://github.com/junan/google-scraper/projects/1#card-56478576

## What happened 👀

User can't register with the same email address by changing uppercase/lowercase letter of email

## Insight 📝

Used Beego's [iexact](https://beego.me/docs/mvc/model/query.md#iexact) operator

## Proof Of Work 📹

Spec should be passed

### CODE REVIEW

File: models/user.go [hunk 1]
```diff
@@ -19,7 +19,10 @@ func CreateUser(u *User) (id int64, err error) {
 
 func (u *User) IsExistingUser() bool {
 	o := orm.NewOrm()
-	err := o.Read(u, "Email")
+	var user User
+
+	err := o.QueryTable(User{}).Filter("email__iexact", u.Email).One(&user)
```
Comment (olivierobert) [inline models/user.go:24]: A different alternative is to use `citext` for the column data type: https://hashrocket.com/blog/posts/working-with-email-addresses-in-postgresql#another-approach. It's usually what I recommend for case insensitive pieces of data such as email.

Comment (junan) [reply models/user.go:24]: I am using Beego's [automatic table generation](https://beego.me/docs/mvc/model/orm.md#registering-model) using `RunSyncdb`. With this I don't need to create any migration to create/update tables. Beego does it automatically by mapping model fields  with database type when the app runs(it does in runtime)

But [Beego](https://beego.me/docs/mvc/model/models.md) does not support `string` to `citext` mapping. For `string` model field type, it only supports `varchar, char, text, json, jsonb`

To use `citext` I need to switch back to manually migration(with `bee` CLI), where I can define `citext` column in raw sql. Also drop automatic table generation code and generate all tables from migrations.

What do you think?
Screen Shot 2021-03-24 at 10 31 05 AM

Comment (olivierobert) [reply models/user.go:24]: Noted. So a custom implementation does makes sense.

**Review by olivierobert — APPROVED** [review]

**Review by gutakk — APPROVED** [review]
I think the PR title is incorrect here. Should it be `can` -> `can't`?

File: models/user.go (hunk 1 above)
Comment (Lahphim) [inline models/user.go:24]: @junan By doing this `__iexact`, why don't we try on with lowercase the email before save and query?
The benefit of doing this:
1. clean email in the database
2. we don't have to suffix with `__iexact` 🤣

Or the other way is set it on migration file as @carryall did 🤣

Comment (junan) [reply models/user.go:24]: Regarding store the email in lowercase, In this [answer](https://stackoverflow.com/a/1503679), it suggested to store as it it.

> Or the other way is set it on migration file as @carryall did

I have  beenusing Beego automatic table generation with `RunSyncdb ` and I don't want to switch back to manual migration where I need create all the table from migration again 😄 

For me using Beego's `__iexact` is ok. What do you think?

**Review by Lahphim — APPROVED** [review]

**Pull request merged**